File: src/com/lin/L2300.java
package src.com.lin;

import java.lang.reflect.Array;
import java.util.Arrays;

/**
 * L2300
 */
public class L2300 {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        int ans[] = new int[spells.length];
        Arrays.sort(potions);
        for (int i = 0; i < spells.length; i++) {
            ans[i] = solve(spells[i], potions, success);
        }
        return ans;
    }
    public int solve(int value,int[] potions, long success) {
        int l = 0;
        int r = potions.length;
        while(l < r) {
            int mid = (l + r) >> 1;
            if(value * 1l * potions[mid] >= success)r = mid;
            else l = mid + 1; 
        }
        return potions.length - r;
    }
}

File: rust/src/l_2300.rs
use crate::Solution;

impl Solution {
    pub fn successful_pairs(spells: Vec<i32>, mut potions: Vec<i32>, success: i64) -> Vec<i32> {
        potions.sort();
        let mut ans: Vec<i32> = vec![0; spells.len()];
        for i in 0..spells.len() {
            let mut l = 0;
            let mut r = potions.len();
            while l < r {
                let mid = (l + r) >> 1;
                if spells[i] as i64 * potions[mid] as i64 >= success {
                    r = mid;
                } else {
                    l = mid + 1;
                }
            }
            ans[i] += if l != potions.len() {
                potions.len() - l
            } else {
                0
            } as i32
        }
        vec![]
    }
}
